### ISSUE TITLE
Query with booleans displaying wikicode breaks preview button

### ISSUE THREAD

**Issue by Malvineous** [SemanticMediaWiki/SemanticMediaWiki#1464]
The manual says you can set an `#ask` query to show different values for boolean types, like this:

```
| ?Field#Yes,No
```

However if you try to do something fancy, like this:

```
| ?Field#<span style="color: green; font-size: 120%;">&#10003;</span>,<span style="color: #AA0000; font-size: 120%;">&#10005;</span>
```

Then the query output is replaced with a large block of HTML code, and the preview button no longer works making it difficult to correct your mistake.

Putting the offending wikicode into a template and doing this produces the same result:

```
| ?Field#{{Tick}},{{Cross}}
```

Is it possible to make the query accept markup code?  If it has to be a template that's fine.  At the very least hopefully it's possible to stop it from breaking the edit window, since the preview button no longer works after this happens.

**Comment by Malvineous** [comment]
Example demonstrating problem on sandbox: http://sandbox.semantic-mediawiki.org/wiki/Issue/1464_%28Boolean_query_values%29

**Comment by mwjames** [comment]
> Is it possible to make the query accept markup code?

See [0] where `=` [1] is used to declare a different label. `#` indicates a formatting rule.

[0] http://sandbox.semantic-mediawiki.org/wiki/Issue/1464_%28Boolean_query_values%29/Label_declaration

[1] https://www.semantic-mediawiki.org/wiki/Help:Displaying_information#Printing_property_values

**Comment by Malvineous** [comment]
That only seems to change the column headings though, whereas I'm wanting to print a green tick instead of the word "True" and a red cross instead of the word "False", in each row of the table.

**Comment by mwjames** [comment]
First of all, the type has to be declared as "Boolean" and if it is a boolean type then you can format the value with the help of `#` such as `#x,-` (or `#✓,✕`) as seen in [0].

If you wanted to support something like `#&#10003;,&#10005;` then you would have to change the code.

[0] http://sandbox.semantic-mediawiki.org/wiki/Issue/1464_%28Boolean_query_values%29/Label_declaration

**Comment by mwjames** [comment]
See also [0].

[0] https://www.semantic-mediawiki.org/wiki/Help:Type_Boolean#Displaying_Boolean_values_in_inline_queries

**Comment by Malvineous** [comment]
Thanks, this is what I have done on the wiki I am having a problem with.  I just forgot to set the type to boolean on the sandbox.  I have updated the sandbox page to demonstrate, and the query output is broken again (I only updated the first query.)  I am happy using `✓` and `✕` but I can't seem to get them to change colour.

**Comment by mwjames** [comment]
> If you wanted to support something like `#&#10003;,&#10005;` then you would have to change the code.

As I said, something like `"#<span style="color: green;">✓</span>,<span style="color: red;">✕</span>" ` will not work for the cited reasons (`=` has a special meaning for the parser which is the cause of the split up html code).

**Comment by Malvineous** [comment]
Oh right, so the `=` in the `<span>` is causing the parser to switch to treating the rest of the line as the label?  That makes sense.  I see the same problem if I put the code into a template instead, so is it practical to modify the code to expand templates later in the process, or would that break other functionality?

**Comment by mwjames** [comment]
> if I put the code into a template instead, so is it practical to modify the code to expand templates later in the process,

It needs to happen before and just for the fun of it you would need to do something like (PS:__I did not test this extensively nor did I run it against our test suite__):

```
{{#ask: [[Has boolean::+]]
 |?Has boolean#<span style="color: green; font-size: 120%;">&#10003;</span>,<span style="color: #AA0000; font-size: 120%;">&#10005;</span>=Label on (&#10003;,&#10005;)
}}
```

```
@@ -358,11 +358,22 @@ class SMWQueryProcessor {
 	 * @since 1.8
 	 */
 	static protected function getSMWPrintRequestFromString( $printRequestString, $showMode ) {
 		global $wgContLang;
 
+		// Temporary encode "=" within an html entity
+		$printRequestString = preg_replace_callback( "/(<(.*?)>(.*?)>)/u", function( $matches ) {
+			foreach ( $matches as $match ) {
+				return str_replace( array( '=' ), array( '-3D' ), $match );
+			}
+		}, $printRequestString );
+
 		$parts = explode( '=', $printRequestString, 2 );
+
+		// Restore temporary encoding
+		$parts[0] = str_replace( array( '-3D' ), array( '=' ), $parts[0] );
```
```
@ -75,14 +75,15 @@ class SMWBoolValue extends SMWDataValue {
 			$this->m_truecaption = '<span style="font-family: sans-serif; ">X</span>';
 			$this->m_falsecaption = '&nbsp;';
 		} else { // format "truelabel, falselabel" (hopefully)
 			$captions = explode( ',', $formatstring, 2 );
 			if ( count( $captions ) == 2 ) { // note: escaping needed to be safe; MW-sanitising would be an alternative
-				$this->m_truecaption = htmlspecialchars( trim( $captions[0] ) );
-				$this->m_falsecaption = htmlspecialchars( trim( $captions[1] ) );
+				$this->m_truecaption = htmlspecialchars_decode( htmlspecialchars( trim( $captions[0] ) ) );
+				$this->m_falsecaption = htmlspecialchars_decode( htmlspecialchars( trim( $captions[1] ) ) );
```

![image](https://cloud.githubusercontent.com/assets/1245473/13904479/5b7ab236-eea2-11e5-87c0-c3242a6dc0d3.png)

**Comment by mwjames** [comment]
Solved by #1518
